src/dataflow/src/source/postgres.rs

```diff
@@ -336,13 +367,36 @@ impl PostgresSourceReader {
 impl SimpleSource for PostgresSourceReader {
     /// The top-level control of the state machine and retry logic
     async fn start(mut self, timestamper: &Timestamper) -> Result<(), SourceError> {
-        // The initial snapshot has no easy way of retrying it in case of connection failures
-        self.produce_snapshot(timestamper)
-            .await
-            .map_err(|e| SourceError {
+        // Buffer rows from snapshot to retract and retry, if initial snapshot fails.
+        // Postgres sources cannot proceed without a successful snapshot.
+        let mut failures: i32 = 0;
+        loop {
+            let mut buffer = NamedTempFile::new().map_err(|e| SourceError {
                 source_name: self.source_name.clone(),
-                error: SourceErrorDetails::Initialization(e.to_string()),
+                error: SourceErrorDetails::FileIO(e.to_string()),
             })?;
+            match self.produce_snapshot(timestamper, &mut buffer).await {
+                Ok(_) => break,
+                Err(e) => {
+                    if failures == 3 {
+                        return Err(SourceError {
+                            source_name: self.source_name.clone(),
+                            error: SourceErrorDetails::Initialization(e.to_string()),
+                        });
+                    }
+
+                    failures += 1;
+                    log::info!("failed to take snapshot: {}", e.to_string());
```

Comment:
There are a couple of things around this error message:

- log level of INFO is not appropriate, this should be a WARN 
- the message does not say that the operation will be retried 
- the message message is missing the name of the source whose snapshot has failed:
- there is no corresponding message when the snapshot is ultimately sucessfull. 

This is the entire set of messages that are printed around a snapshotting failure:

```
2021-05-23T10:39:46.337153Z  INFO coord::catalog: create source materialize.public.mz_source (u1)    
2021-05-23T10:39:46.337216Z  INFO coord::catalog: create index materialize.public.mz_source_primary_idx (u2)    
2021-05-23T10:39:46.341401Z  INFO coord::timestamp: Timestamping Source u1 with Real Time Consistency.    
2021-05-23T10:39:46.393222Z  INFO coord::catalog: create view materialize.public.ten (u3)    
2021-05-23T10:39:46.393258Z  INFO coord::catalog: create view materialize.public.t1 (u4)    
2021-05-23T10:39:46.569242Z  INFO dataflow::source::postgres: failed to take snapshot: connection closed 
```

and that is all there is, all INFO lines.

Reply:
I think I've addressed this comment fully in my newest commit, but happy to make further changes if the messages are still unclear!